### ISSUE TITLE
Teach data sanitization in JS projects

### ISSUE THREAD

**Issue by moT01** [freeCodeCamp/freeCodeCamp#54547]
### Describe the Issue

In the [authors page project](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/#learn-fetch-and-promises-by-building-an-fcc-authors-page) and the [forum leaderboard project](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/#learn-asynchronous-programming-by-building-an-fcc-forum-leaderboard), we are having users fetch data from an external resource and inserting it into the DOM via innerHTML. This isn't a good practice as the data could theoretically be malicious. It's not because we control the data, but we should change these two projects to have users sanitize the data.

### Affected Page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/

**Comment by a2937** [comment]
So for the authors page project, we probably need to add a step where step 16 is currently at that deals with sanitizing the data. I think we should consider adding some innocent but demonstrative bad data into our API. Stuff like special characters before it can actually be displayed on the page. At minimum, adding some less than and greater than symbols that need encoded could help. 

As for forum leaderboard project, roughly around step 11 is where we should do the same thing.

**Comment by ShaunSHamilton** [comment]
Because we are not exactly set up for modules, we might want to go the route of:

> You **could** insert the data into your element by setting the `innertHTML` property. However, this exposes your page to _____. Instead, use the `innerText` property to insert the data: ....

**Comment by jdwilkin4** [comment]
> Instead, use the innerText property to insert the data: 

innerText doesn't work though

Are you suggesting to make this change?

```js
   authorContainer.innerText += `
    <div id="${index}" class="user-card">
      <h2 class="author-name">${author}</h2>
      <img class="user-img" src="${image}" alt="${author} avatar">
      <div class="purple-divider"></div>
      <p class="bio">${bio.length > 50 ? bio.slice(0, 50) + '...' : bio}</p>
      <a class="author-link" href="${url}" target="_blank">${author} author page</a>
    </div>
  `;
```
because if so, it would produce this result

<img width="617" alt="Screenshot 2024-04-29 at 6 22 01 AM" src="https://github.com/freeCodeCamp/freeCodeCamp/assets/67210629/9356ce90-d318-46f5-bb7e-814dc31b4d2c">

**Comment by ShaunSHamilton** [comment]
> Are you suggesting to make this change?

No, I suggest splitting the stuff we control and the text into different parts.

I see why that might not work very easily; Previously, I had just looked at the forum leaderboard:

```js
const showLatestPosts = (data) => {
  const { topic_list, users } = data;
  const { topics } = topic_list;

  postsContainer.innerHTML = topics.map((item) => {
    const {
      id,
      title,
      views,
      posts_count,
      slug,
      posters,
      category_id,
      bumped_at,
    } = item;

    return `
    <tr>
      <td>
        <p class="post-title">${title}</p>

        ${forumCategory(category_id)}
      </td>
      <td>
        <div class="avatar-container">
          ${avatars(posters, users)}
        </div>
      </td>
      <td>${posts_count - 1}</td>
      <td>${viewCount(views)}</td>
      <td>${timeAgo(bumped_at)}</td>
    </tr>`;
  }).join("");
};
```

**Comment by jdwilkin4** [comment]
> No, I suggest splitting the stuff we control and the text into different parts.

That might work.

We technically control all of the content for the authors page project since it is just a hardcoded json file
https://github.com/freeCodeCamp/cdn/blob/main/build/curriculum/news-author-page/authors.json

**Comment by ojeytonwilliams** [comment]
@moT01 @jdwilkin4 does the upcoming curriculum cover/plan to cover this?

**Comment by jdwilkin4** [comment]
@ojeytonwilliams 

The new cert does talk about the security concerns around using innerHTML but we don't cover data sanitization specifically.
We could add a lecture video for that, but I am not sure where the best place for that would be.

**Comment by ojeytonwilliams** [comment]
Okay, thanks @jdwilkin4. In that case, I think it's fine to leave this issue up to remind ourselves to cover it when we have the chance.

**Comment by moT01** [comment]
Closing as stale.
